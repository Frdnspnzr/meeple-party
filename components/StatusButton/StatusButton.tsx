import useCollectionStatus from "@/hooks/api/useCollectionStatus";
import classNames from "classnames";
import { useCallback, useMemo } from "react";
import Spinner from "../Spinner/Spinner";
import IconCollectionOwn from "../icons/CollectionOwn";
import IconCollectionWantToPlay from "../icons/CollectionWantToPlay";
import IconCollectionWishlist from "../icons/CollectionWishlist";
import styles from "./statusbutton.module.css";

interface StatusButtonProps extends React.HTMLAttributes<HTMLDivElement> {
  gameId: number;
  status: "own" | "wanttoplay" | "wishlist";
}

export default function StatusButton({
  status,
  gameId,
  ...props
}: Readonly<StatusButtonProps>) {
  const { data, isLoading, mutate } = useCollectionStatus(gameId);
  const icon = useMemo(() => {
    if (status === "own") {
      return <IconCollectionOwn />;
    } else if (status === "wanttoplay") {
      return <IconCollectionWantToPlay />;
    } else if (status === "wishlist") {
      return <IconCollectionWishlist />;
    }
  }, [status]);
  const activeColor = useMemo(() => {
    if (status === "own") {
      return "var(--color-collection-own)";
    } else if (status === "wanttoplay") {
      return "var(--color-collection-want-to-play)";
    } else if (status === "wishlist") {
      return "var(--color-collection-wishlist)";
    }
  }, [status]);
  const active = useMemo(() => {
    if (!data) {
      return false;
    } else if (status === "own") {
      return data.own;
    } else if (status === "wishlist") {
      return data.wishlist;
    } else if (status === "wanttoplay") {
      return data.wantToPlay;
    }
  }, [data, status]);
  const onClick = useCallback(() => {
    if (!data) {
      return;
    } else if (status === "own") {
      mutate({ ...data, own: !data.own });
    } else if (status === "wanttoplay") {
      mutate({ ...data, wantToPlay: !data.wantToPlay });
    } else if (status === "wishlist") {
      mutate({ ...data, wishlist: !data.wishlist });
    }
  }, [status, mutate, data]);

  if (isLoading) {
    return (
      <Spinner
        {...props}
        className={classNames(styles.spinner, props.className)}
      />
    );
  } else {
    return (
      <div
        {...props}
        onClick={onClick}
        className={classNames(styles.button, props.className)}
        style={{
          ...props.style,
          color: active ? activeColor : "unset",
        }}
      >
        {icon}
      </div>
    );
  }
}
