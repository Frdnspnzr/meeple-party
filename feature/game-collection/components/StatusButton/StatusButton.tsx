import useCollectionStatus from "@/feature/game-collection/hooks/useCollectionStatus";
import classNames from "classnames";
import { useCallback, useMemo } from "react";
import Spinner from "../../../../lib/components/Spinner/Spinner";
import IconCollectionOwn from "../../../../lib/icons/CollectionOwn";
import IconCollectionWantToPlay from "../../../../lib/icons/CollectionWantToPlay";
import IconCollectionWishlist from "../../../../lib/icons/CollectionWishlist";
import styles from "./statusbutton.module.css";

interface StatusButtonProps extends React.HTMLAttributes<HTMLDivElement> {
  gameId: number;
  status: "own" | "wanttoplay" | "wishlist";
}

export default function StatusButton({
  status,
  gameId,
  ...props
}: Readonly<StatusButtonProps>) {
  const { data, isLoading, mutate } = useCollectionStatus(gameId);
  const icon = useMemo(() => {
    if (status === "own") {
      return <IconCollectionOwn />;
    } else if (status === "wanttoplay") {
      return <IconCollectionWantToPlay />;
    } else if (status === "wishlist") {
      return <IconCollectionWishlist />;
    }
  }, [status]);
  const activeColor = useMemo(() => {
    if (status === "own") {
      return "var(--color-collection-own)";
    } else if (status === "wanttoplay") {
      return "var(--color-collection-want-to-play)";
    } else if (status === "wishlist") {
      return "var(--color-collection-wishlist)";
    }
  }, [status]);
  const active = useMemo(() => {
    if (!data) {
      return false;
    } else if (status === "own") {
      return data.own;
    } else if (status === "wishlist") {
      return data.wishlist;
    } else if (status === "wanttoplay") {
      return data.wantToPlay;
    }
  }, [data, status]);
  const updateStatus = useCallback(() => {
    if (!data) {
      return;
    } else if (status === "own") {
      mutate({ ...data, own: !data.own });
    } else if (status === "wanttoplay") {
      mutate({ ...data, wantToPlay: !data.wantToPlay });
    } else if (status === "wishlist") {
      mutate({ ...data, wishlist: !data.wishlist });
    }
  }, [status, mutate, data]);

  if (isLoading) {
    return (
      <Spinner
        {...props}
        className={classNames(styles.spinner, props.className)}
      />
    );
  } else {
    return (
      <div
        {...props}
        onClick={updateStatus}
        onKeyDown={(e) => {
          if (e.key === "Enter") {
            updateStatus();
          }
        }}
        tabIndex={0}
        className={classNames(styles.button, props.className)}
        style={{
          ...props.style,
          color: active ? activeColor : "unset",
        }}
        role="button"
      >
        {icon}
      </div>
    );
  }
}
