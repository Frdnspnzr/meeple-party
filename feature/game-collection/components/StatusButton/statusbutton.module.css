.button {
  font-size: 1.75rem;
}

.spinner {
  margin: auto;
}

.text {
  font-size: 0.7rem;
  text-align: center;
  font-weight: lighter;
}
