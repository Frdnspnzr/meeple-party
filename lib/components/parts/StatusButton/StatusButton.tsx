import Spinner from "@/components/Spinner/Spinner";
import IconCollectionOwn from "@/components/icons/CollectionOwn";
import IconCollectionWantToPlay from "@/components/icons/CollectionWantToPlay";
import IconCollectionWishlist from "@/components/icons/CollectionWishlist";
import { useTranslation } from "@/i18n/client";
import classNames from "classnames";
import { useMemo } from "react";
import styles from "./statusbutton.module.css";

interface StatusButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
  status: "own" | "wanttoplay" | "wishlist";
  loading?: boolean;
  active: boolean;
  toggle?: () => void;
}

export default function StatusButton({
  status,
  loading = false,
  active,
  toggle = () => {},
  ...props
}: Readonly<StatusButtonProps>) {
  const { t } = useTranslation();
  const icon = useMemo(() => {
    if (status === "own") {
      return <IconCollectionOwn />;
    } else if (status === "wanttoplay") {
      return <IconCollectionWantToPlay />;
    } else if (status === "wishlist") {
      return <IconCollectionWishlist />;
    }
  }, [status]);
  const activeColor = useMemo(() => {
    if (status === "own") {
      return "var(--color-collection-own)";
    } else if (status === "wanttoplay") {
      return "var(--color-collection-want-to-play)";
    } else if (status === "wishlist") {
      return "var(--color-collection-wishlist)";
    }
  }, [status]);
  const label = useMemo(() => {
    if (status === "own") {
      return t("States.Own");
    } else if (status === "wanttoplay") {
      return t("States.WantToPlay");
    } else if (status === "wishlist") {
      return t("States.Wishlist");
    }
  }, [status, t]);

  if (loading) {
    return (
      <Spinner
        // {...props}
        size="small"
        className={classNames(styles.spinner, props.className)}
      />
    );
  } else {
    return (
      <button
        {...props}
        onClick={(e) => {
          if (props.onClick) {
            props.onClick(e);
          }
          toggle();
        }}
        onKeyDown={(e) => {
          if (props.onKeyDown) {
            props.onKeyDown(e);
          }
          if (e.key === "Enter") {
            toggle();
          }
        }}
        tabIndex={0}
        className={classNames(styles.button, props.className)}
        style={{
          ...props.style,
          color: active ? activeColor : "unset",
        }}
        role="button"
        aria-label={label}
      >
        {icon}
      </button>
    );
  }
}
