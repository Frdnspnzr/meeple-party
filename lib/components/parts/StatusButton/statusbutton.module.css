.button {
  text-align: center;
}

.spinner {
  margin: auto;
}