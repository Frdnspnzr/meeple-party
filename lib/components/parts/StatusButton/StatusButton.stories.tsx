import type { Meta, StoryObj } from "@storybook/react";
import StatusButton from "./StatusButton";

const meta = {
  title: "Parts/StatusButton",
  component: StatusButton,
  tags: ["autodocs"],
} satisfies Meta<typeof StatusButton>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Own: Story = {
  args: {
    status: "own",
    active: true,
  },
};

export const WantToPlay: Story = {
  args: {
    status: "wanttoplay",
    active: true,
  },
};

export const Wishlist: Story = {
  args: {
    status: "wishlist",
    active: true,
  },
};

export const Loading: Story = {
  args: {
    status: "own",
    active: true,
    loading: true,
  },
};
