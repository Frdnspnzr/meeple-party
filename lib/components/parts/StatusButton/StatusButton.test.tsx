import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import StatusButton from "./StatusButton";

jest.mock("@/i18n/client");
jest.mock("@/lib/hooks/useCollectionStatus");

describe("Status Button", () => {
  test.each(["own", "wanttoplay", "wishlist"])(
    "matches snapshot for %s when active",
    async (status) => {
      const { container } = render(
        <StatusButton status={status as any} active={true} />
      );
      expect(container).toMatchSnapshot();
    }
  );
  test.each(["own", "wanttoplay", "wishlist"])(
    "matches snapshot for %s when not active",
    async (status) => {
      const { container } = render(
        <StatusButton status={status as any} active={false} />
      );
      expect(container).toMatchSnapshot();
    }
  );
  it("matches snapshot when loading", async () => {
    const { container } = render(
      <StatusButton status="own" active={false} loading={true} />
    );
    expect(container).toMatchSnapshot();
  });
  it("passes class names to container", async () => {
    const { container } = render(
      <StatusButton status="own" active={true} className="passed class names" />
    );
    expect(container).toMatchSnapshot();
  });
  it("passes styles to container", async () => {
    const { container } = render(
      <StatusButton
        status="own"
        active={true}
        style={{ width: "100%", alignSelf: "end", marginBlockStart: "1em" }}
      />
    );
    expect(container).toMatchSnapshot();
  });
  test.each(["own", "wanttoplay", "wishlist"])(
    "calls callback on clicking button for state %s",
    async (status) => {
      const user = userEvent.setup();
      let called = false;
      const callback = () => {
        called = true;
      };
      render(
        <StatusButton status={status as any} active={false} toggle={callback} />
      );
      const button = screen.getByRole("button");
      await user.click(button);
      expect(called).toBeTruthy();
      expect(button).toHaveAccessibleName(
        status === "own"
          ? "States.Own"
          : status === "wanttoplay"
            ? "States.WantToPlay"
            : "States.Wishlist"
      );
    }
  );
  it("calls callback on key press", async () => {
    const user = userEvent.setup();
    let called = false;
    const callback = () => {
      called = true;
    };
    render(<StatusButton status="own" active={false} toggle={callback} />);
    const button = screen.getByRole("button");
    button.focus();
    await user.keyboard("[Enter]");
    expect(called).toBeTruthy();
  });
  it("calls original event handler on clicking button", async () => {
    const user = userEvent.setup();
    let called = false;
    const callback = () => {
      called = true;
    };
    render(<StatusButton status="own" active={false} onClick={callback} />);
    const button = screen.getByRole("button");
    await user.click(button);
    expect(called).toBeTruthy();
  });
  it("calls original event handler on key press", async () => {
    const user = userEvent.setup();
    let called = false;
    const callback = () => {
      called = true;
    };
    render(<StatusButton status="own" active={false} onKeyDown={callback} />);
    const button = screen.getByRole("button");
    button.focus();
    await user.keyboard("[Enter]");
    expect(called).toBeTruthy();
  });
});
