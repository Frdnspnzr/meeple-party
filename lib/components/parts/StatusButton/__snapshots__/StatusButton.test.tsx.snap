// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status Button matches snapshot for own when active 1`] = `
<div>
  <button
    aria-label="States.Own"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-box-seam-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot for own when not active 1`] = `
<div>
  <button
    aria-label="States.Own"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-box-seam-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot for wanttoplay when active 1`] = `
<div>
  <button
    aria-label="States.WantToPlay"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-dice-3-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot for wanttoplay when not active 1`] = `
<div>
  <button
    aria-label="States.WantToPlay"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-dice-3-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot for wishlist when active 1`] = `
<div>
  <button
    aria-label="States.Wishlist"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-gift-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot for wishlist when not active 1`] = `
<div>
  <button
    aria-label="States.Wishlist"
    class="button"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-gift-fill"
    />
  </button>
</div>
`;

exports[`Status Button matches snapshot when loading 1`] = `
<div>
  <div
    class="spinner spinner-border spinner-border-sm"
  />
</div>
`;

exports[`Status Button passes class names to container 1`] = `
<div>
  <button
    aria-label="States.Own"
    class="button passed class names"
    role="button"
    tabindex="0"
  >
    <i
      class="bi bi-box-seam-fill"
    />
  </button>
</div>
`;

exports[`Status Button passes styles to container 1`] = `
<div>
  <button
    aria-label="States.Own"
    class="button"
    role="button"
    style="width: 100%; align-self: end; margin-block-start: 1em;"
    tabindex="0"
  >
    <i
      class="bi bi-box-seam-fill"
    />
  </button>
</div>
`;
